cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+  The difference between equality and equivalence in Cypher today is small and subtle, and limited to testing two instances of the value `null` to each other.
+
+    ** In equality, `null = null` is `null`.
+    ** In equivalence, used by grouping and `DISTINCT`, instances of `null` are always treated as being the same value.
+    ** However, equality treats `null` values differently if they are an element of a list or a map value.
+    ** Similar rules apply for `NaN`.
+
+== Proposal
+
+=== Comparability
+
+We propose that comparability be defined between any pair of values, as specified below (Please note that `null` is interpreted as meaning "unknown" or "could be any value" in the context of comparability):
+
+* Integers are compared numerically.
+* Floats (excluding `NaN` and the Infinities) are compared numerically.
+* Arbitrary numbers (excluding `NaN` and the Infinities) are compared to each other numerically, i.e. the comparison behaves as if both numbers would have been coerced to arbitrary precision big decimals (currently outside the Cypher type system) before comparing them with each other numerically.
+* Positive infinity is of type `FLOAT`, equal to itself and larger than any other number (excluding `NaN`).
+* Negative infinity is of type `FLOAT`, equal to itself and smaller than any other number (excluding `NaN`).
+* Comparing a number to a `NaN` always yields `null`.
+* Booleans are compared such that `false < true`.
+* Strings are compared in dictionary order, i.e. characters are compared pairwise in order and characters missing in a shorter string are considered to be smaller than any other character. For example, `'a' < 'aa'`.
+* Lists are compared in dictionary order, i.e. list elements are compared pairwise in order and elements missing in a shorter list are considered to be smaller than any other value. For example, `[1] < [1,0]`.
+* Lists containing `null` values are incomparable, i.e. comparing them with any other value yields `null`.
+* Maps are compared as if they where lists of lists, where the inner lists are the key-value pairs of the maps, ordered in ascending sort order for the keys. Comparing a map to a value that is not a map always yields `null`.
+* Maps containing explicit `null` value entries are incomparable, i.e. comparing them with any other value yields `null`.
+* The comparison order for nodes (and relationships resp.) is undefined, and an implementation may choose an order for these freely, e.g. by using some kind of id. This order must not change during the execution of the query (i.e. it must be stable in the presence of updates). Comparing a node to a relationship always yields `null`.
+* Paths are compared as if they where a list of alternating nodes and relationships of the path from the start node to the end node. Comparing a path to a value that is not a path always yields `null`.
```

Comment:
The above description does not say, however, precisely how the paths are compared. Is it numeric or dictionary order of the composite ids? What about length? Examples would be very beneficial too.